### PR TITLE
Issue 384 randomize nodeset seed

### LINKED ISSUE #384: Randomize cannot be used with a nodeset path
#### Software versions 
JavaRosa v2.12, Collect v1.18

#### Problem description
It seems like paths aren't allowed as a seed.

This works:
`<itemset nodeset="randomize(instance('opt_abcd')/root/item, 12345)">`

This fails:
`<itemset nodeset="randomize(instance('opt_abcd')/root/item, /rand/my_seed)">`

The error message I get is
```
java.lang.NumberFormatException: For input string: "/rand/my_seed"
	at java.lang.Long.parseLong(Long.java:432)
	at java.lang.Long.parseLong(Long.java:485)
	at org.javarosa.xform.parse.RandomizeHelper.parseSeed(RandomizeHelper.java:44)
	at org.javarosa.xform.parse.XFormParser.parseItemset(XFormParser.java:1321)
	at org.javarosa.xform.parse.XFormParser.parseControl(XFormParser.java:1000)
	at org.javarosa.xform.parse.XFormParser.parseControl(XFormParser.java:910)
	at org.javarosa.xform.parse.XFormParser.access$200(XFormParser.java:120)
	at org.javarosa.xform.parse.XFormParser$1$6.handle(XFormParser.java:242)
	at org.javarosa.xform.parse.XFormParser.parseElement(XFormParser.java:531)
	at org.javarosa.xform.parse.XFormParser.parseElement(XFormParser.java:540)
	at org.javarosa.xform.parse.XFormParser.parseElement(XFormParser.java:540)
	at org.javarosa.xform.parse.XFormParser.parseDoc(XFormParser.java:447)
	at org.javarosa.xform.parse.XFormParser.parse(XFormParser.java:358)
	at org.javarosa.xform.util.XFormUtils.getFormFromInputStream(XFormUtils.java:81)
	at org.odk.collect.android.tasks.FormLoaderTask.createFormDefFromCacheOrXml(FormLoaderTask.java:236)
	at org.odk.collect.android.tasks.FormLoaderTask.doInBackground(FormLoaderTask.java:126)
	at org.odk.collect.android.tasks.FormLoaderTask.doInBackground(FormLoaderTask.java:70)
	at android.os.AsyncTask$2.call(AsyncTask.java:304)
	at java.util.concurrent.FutureTask.run(FutureTask.java:237)
	at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:243)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
	at java.lang.Thread.run(Thread.java:761)
```

#### Steps to reproduce the problem
Load [rand.xml.zip](https://github.com/opendatakit/javarosa/files/2596539/rand.xml.zip) into Collect. Open it and it'll crash with the above error.

#### Expected behavior
A seed using the value at /rand/my_seed is loaded.

#### Other information 
This is urgent because it provides a workaround for https://github.com/XLSForm/pyxform/issues/238.

### PR DESCRIPTION
**Pull request by ggalmazor**
Closes #384

#### What has been done to verify that this works as intended?
Added new automated tests

#### Why is this the best possible solution? Were any other approaches considered?
No other approaches have been explored, although there could be more elegant alternatives to the changes in `ItemsetBinding`.

#### Are there any risks to merging this code? If so, what are they?
No. These changes don't break previous tests and all the additions should be (not checked) compatible with the target Android APIs

### CODE REVIEW

File: resources/org/javarosa/xpath/expr/randomize.xml [hunk 1]
```diff
@@ -4,20 +4,24 @@
...
             <instance>
-                <randomize id="randomize">
-                    <fruit1/>
-                    <fruit2/>
-                    <seededFruit1/>
-                    <seededFruit2/>
-                    <randomValue/>
-                    <seededRandomValue/>
+                <data id="randomize">
+                    <no-seed-random-value/>
```
Comment (ggalmazor) [inline resources/org/javarosa/xpath/expr/randomize.xml]: I would have used some groups here to make it even more readable/structured, but then it wouldn't work with the current testing tools (they assume everything is at the root instance level)

File: src/org/javarosa/core/model/ItemsetBinding.java [hunk 2]
```diff
@@ -55,18 +59,27 @@
     private List<SelectChoice> choices; //dynamic choices -- not serialized, obviously
 
     public boolean randomize = false;
-    public Long randomSeed = null;
+    public XPathNumericLiteral randomSeedNumericExpr = null;
```
Comment (dcbriccetti) [inline src/org/javarosa/core/model/ItemsetBinding.java]: Not a show stopper, but aren’t these initializers redundant?

File: test/org/javarosa/xform/parse/RandomizeHelperTest.java [hunk 3]
```diff
@@ -34,15 +34,36 @@
...
-        assertEquals("/some/path",                              cleanNodesetDefinition("randomize( /some/path )"));
-        assertEquals("/some/path",                              cleanNodesetDefinition("randomize(/some/path,33)"));
-        assertEquals("/some/path",                              cleanNodesetDefinition("randomize(/some/path, 33)"));
-        assertEquals("/some/path",                              cleanNodesetDefinition("randomize( /some/path , 33)"));
-        assertEquals("/some/path[someFilter]",                  cleanNodesetDefinition("randomize(/some/path[someFilter])"));
-        assertEquals("/some/path[someFilter]",                  cleanNodesetDefinition("randomize(/some/path[someFilter], 33)"));
-        assertEquals("/some/path[someFilter(with, commas)]",    cleanNodesetDefinition("randomize(/some/path[someFilter(with, commas)])"));
-        assertEquals("/some/path[someFilter(with, commas)]",    cleanNodesetDefinition("randomize(/some/path[someFilter(with, commas)], 33)"));
+        assertEquals("/some/path", cleanNodesetDefinition("randomize(/some/path)"));
```
Comment (dcbriccetti) [inline test/org/javarosa/xform/parse/RandomizeHelperTest.java]: I’l liked the two clear columns, but that’s not for everybody.

File: src/org/javarosa/core/model/ItemsetBinding.java (hunk 2 above)
Comment (ggalmazor) [reply src/org/javarosa/core/model/ItemsetBinding.java]: I guess they look like redundant, although IntelliJ isn't highlighting them for me... Are you confident that we can remove them?

Comment (dcbriccetti) [reply src/org/javarosa/core/model/ItemsetBinding.java]: I, also, like the reassurance of a message from IDEA. I’m too lazy to look up field initialization in the JLS right now, and I’m not gonna go out on a limb. :-)

Comment (ggalmazor) [reply src/org/javarosa/core/model/ItemsetBinding.java]: Let's leave it then :)

**Pull request merged**